File: backend/app/models/visitor.py
from datetime import date

from sqlalchemy import (
    String,
    Boolean,
    Date,
)
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship,
)

from app.extensions import db
from app.models.base import UUIDMixin, TimestampMixin


class Visitor(db.Model, UUIDMixin, TimestampMixin):
    __tablename__ = "visitors"

    first_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    middle_name: Mapped[str | None] = mapped_column(
        String(100),
    )

    last_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    gender: Mapped[str | None] = mapped_column(
        String(20),
    )

    date_of_birth: Mapped[date | None] = mapped_column(
        Date,
    )

    email: Mapped[str | None] = mapped_column(
        String(255),
        index=True,
    )

    phone: Mapped[str] = mapped_column(
        String(30),
        nullable=True,
        index=True,
    )

    company: Mapped[str | None] = mapped_column(
        String(255),
    )

    address: Mapped[str | None] = mapped_column(
        String(255),
    )

    nationality: Mapped[str | None] = mapped_column(
        String(100),
    )

    id_number: Mapped[str | None] = mapped_column(
        String(100),
        unique=True,
        index=True,
    )

    passport_number: Mapped[str | None] = mapped_column(
        String(100),
        unique=True,
        index=True,
    )

    vehicle_registration: Mapped[str | None] = mapped_column(
        String(50),
        index=True,
    )

    photo_url: Mapped[str | None] = mapped_column(
        String(500),
    )

    is_blacklisted: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False,
    )

    notes: Mapped[str | None] = mapped_column(
        String(1000),
    )

    visits = relationship(
    "Visit",
    back_populates="visitor",
    cascade="all, delete-orphan",
    )

    @property
    def full_name(self):
        names = [
            self.first_name,
            self.middle_name,
            self.last_name,
        ]
        return " ".join(filter(None, names))

File: frontend/src/types/visitor.ts
export interface Visitor {
  id: string;
 
  first_name: string;
  middle_name: string | null;
  last_name: string;
 
  gender: string | null;
  date_of_birth: string | null;
 
  phone: string;
  email: string | null;
 
  company: string | null;
  address: string | null;
  nationality: string | null;
 
  id_number: string | null;
  passport_number: string | null;
 
  vehicle_registration: string | null;
 
  photo_url: string | null;
 
  is_blacklisted: boolean;
 
  notes: string | null;
 
  created_at?: string;
  updated_at?: string;
}
 
export interface CreateVisitorPayload {
  first_name: string;
  middle_name?: string | null;
  last_name: string;
 
  gender?: "Male" | "Female" | "Other" | null;
 
  date_of_birth?: string | null;
 
  phone: string;
  email?: string | null;
 
  company?: string | null;
  address?: string | null;
  nationality?: string | null;
 
  id_number?: string | null;
  passport_number?: string | null;
 
  vehicle_registration?: string | null;
 
  photo_url?: string | null;
 
  is_blacklisted?: boolean;
 
  notes?: string | null;
}
 
export type UpdateVisitorPayload =
  Partial<CreateVisitorPayload>;